### PR TITLE
feat(manager/sbt): Implement %%% dependency specifiers

### LINKED ISSUE #19650: sbt: Support %%% dependencies (e.g. Scala.js)
### What would you like Renovate to be able to do?

When specifying a sbt dependency with
```scala
libraryDependencies ++= Seq(
    "com.lihaoyi" %%% "utest" % "0.8.1" % Test,
)
```
renovate does not detect it. This is syntax that is used by e.g. https://www.scala-js.org/doc/project/dependencies.html

### If you have any ideas on how this should be implemented, please tell us here.

A test would need to be added to https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/sbt/__fixtures__/sample.sbt I think and then %%% should probably behave like %% as that's what happens on the JVM.

Also seems like you would need to add another option at https://github.com/renovatebot/renovate/blob/36e7cb6e6909f27e26e98538fc42e5fcb048f24b/lib/modules/manager/sbt/extract.ts#L203 which looks straightforward at first sight.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by mohe2015**
## Changes

Dependencies that use the `%%%` specifier to make them work with Scala.js and Scala on the JVM are updateable now.

Note that this doesn't work for Scala.js only dependencies as you can see on https://github.com/mohe2015/scalajs-tutorial/pull/1:
`Failed to look up sbt-package dependency org.scala-js:scalajs-dom`

## Context

- Closes partially #19650 (Except Scala.js-only dependencies)

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (https://github.com/mohe2015/scalajs-tutorial)

### CODE REVIEW

File: lib/modules/manager/sbt/scala.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/*
```
Comment (zharinov) [inline lib/modules/manager/sbt/scala.ts:1]: Need to bump `good-enough-parser` dependency to `1.1.22` which contains the change you need

Comment (viceice) [reply lib/modules/manager/sbt/scala.ts:1]: will trigger update via dashboard

Comment (viceice) [reply lib/modules/manager/sbt/scala.ts:1]: - #19672

**Comment by zharinov** [comment]
Probably we can also make `scalajs-dom` package being updated too 🤔
Though, it's the problem of [datasource](https://github.com/renovatebot/renovate/blob/cc35a6a7a7148af334e5486c313718535fb060f0/lib/modules/datasource/sbt-package/index.ts#L51) not parser. I can't remember why we filter those directories.

**Comment by mohe2015** [comment]
> Probably we can also make `scalajs-dom` package being updated too thinking Though, it's the problem of [datasource](https://github.com/renovatebot/renovate/blob/cc35a6a7a7148af334e5486c313718535fb060f0/lib/modules/datasource/sbt-package/index.ts#L51) not parser. I can't remember why we filter those directories.

Should be possible, I'm not interested in implementing that though.


I updated the PR with the new version of good-enough-parser. Thanks for fixing it there!

**Comment by mohe2015** [comment]
Also why do I always get the following diff
```diff
diff --git a/yarn.lock b/yarn.lock
index 5697e1083..1b2a50372 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -2449,7 +2449,6 @@
 
 "@renovate/eslint-plugin@https://github.com/renovatebot/eslint-plugin#v0.0.5":
   version "0.0.5"
-  uid "1923e2e2776549cb25a2c6ea09915ee31bff6d05"
   resolved "https://github.com/renovatebot/eslint-plugin#1923e2e2776549cb25a2c6ea09915ee31bff6d05"
 
 "@renovatebot/osv-offline-db@1.0.1":
```
when installing? Is that something that should be updated in `main`?

**Comment by zharinov** [comment]
No, this is our github-hosted small package for which we haven't decide how to publish it

**Comment by zharinov** [comment]
So, current solution is please don't commit `yarn.lock` :)

**Review by viceice — APPROVED** [review]

**Comment by zharinov** [comment]
> Should be possible, I'm not interested in implementing that though.

But are you interested in having it working? I may take a look...

**Comment by mohe2015** [comment]
> > Should be possible, I'm not interested in implementing that though.
> 
> But are you interested in having it working? I may take a look...

Definitely am interested in having it working.

**Review by zharinov — APPROVED** [review]

**Comment by mohe2015** [comment]
Of course this broke auto-merge...

But are you requiring an updated base branch in the PRs? And does that mean you would need a merge-train like feature to make this work better?

**Comment by zharinov** [comment]
> > > Should be possible, I'm not interested in implementing that though.
> > 
> > 
> > But are you interested in having it working? I may take a look...
> 
> Definitely am interested in having it working.

Quick update. It seems like available version range for every `sjs` library depends on particular `scala-js` version, and available version range for `scala-js` depends on particular version of `sbt`, and available `sbt` version range depends on particular `scala` version.

We managed to make it work for the packages for particular `scala` version, but unfortunately it seems like it isn't our priority to handle the general case. Of course, contributions are welcome from anyone interested.

**Comment by mohe2015** [comment]
> [...]
> We managed to make it work for the packages for particular `scala` version, but unfortunately it seems like it isn't our priority to handle the general case. Of course, contributions are welcome from anyone interested.

Do you have a branch with your wip work? Also would it be so bad to add a method that is not always working but would fall back to a no-op?

**Comment by zharinov** [comment]
I have no WIP, but the outline of what needs to be done is something like that:

Extract level:
1. Extract `scalaVersion` (done), `sbt` version and `scalajs`/`scala-native` versions
2. Provide these versions as the auxilary information to every package extracted

Datasource level:
1. Based on the versions provided, determine which index subdirectories we need to follow. E.g., for `scalajs-dom`, it can be [`scalajs-dom_sjs1_2.11`](https://repo.maven.apache.org/maven2/org/scala-js/scalajs-dom_sjs1_2.11/), [`scalajs-dom_sjs1_2.12`](https://repo.maven.apache.org/maven2/org/scala-js/scalajs-dom_sjs1_2.12/), [`scalajs-dom_sjs1_2.13`](https://repo.maven.apache.org/maven2/org/scala-js/scalajs-dom_sjs1_2.13/), or [`scalajs-dom_sjs1_3`](https://repo.maven.apache.org/maven2/org/scala-js/scalajs-dom_sjs1_3/).

Actually, I just realized it's maybe not that hard: for `scalajs-dom` we additionally need to know `scalajs` plugin version. To update `scalajs` plugin itself, we need to know `sbt` version, but this is the problem orthogonal to `scalajs-dom` update.

Version matching can be a bit tricky, since we have full `scalajs` version while directories can use just major part. Though, it's quite doable.

Well, seems like we need to create new issue based on this discussion.

**Pull request merged**